Improve OCDB Time Machine

dberzano opened the pull request:
* Exit with `AliFatal` in case `SetMaxDate()` is called after the run was set already
* Support setting the date for individual storages: individual storage settings will override global ones

**Note:** OCDB Time Machine feature introduced in #465, with a bug fix in #477.

alibuild commented:
f2e615a3d8917656883a5dbf76a57ee5d042e799: approval required: 1 of @jgrosseo (Jan Fiete Grosse-Oetringhaus), @chiarazampolli (Chiara Zampolli), @qgp (Jochen Klein), @dberzano (Dario Berzano), @sawenzel (Sandro Christian Wenzel), @ktf (Giulio Eulisse), @shahor02 (Ruben Shahoyan)

_Comment with `+1` to approve and allow automatic merging,or with `+test` to run tests only. **Please comment on the pull request: [click here](https://github.com/alisw/AliRoot/pull/486) and comment at the bottom of the page.**_

dberzano commented:
+test

alibuild commented:
f2e615a3d8917656883a5dbf76a57ee5d042e799: testing approved: will not be automatically merged; starting testing. If testing succeeds, merging will require further approval from 1 of @jgrosseo (Jan Fiete Grosse-Oetringhaus), @chiarazampolli (Chiara Zampolli), @qgp (Jochen Klein), @dberzano (Dario Berzano), @sawenzel (Sandro Christian Wenzel), @ktf (Giulio Eulisse), @shahor02 (Ruben Shahoyan)

dberzano commented:
@shahor02: let's come back to the examples you have sent me via email and comment. This new PR should support all of them.

### 1: standard call (OK)
```
man = AliCDBManager::Instance();
man->SetDefaultStorage("raw://");
man->SetRun(281750);
man->Get("GRP/CTP/Aliases");
```

Returns `v3` as expected (it's the latest).

### 2: MaxDate after run number (OK, fixed)
```
man = AliCDBManager::Instance();
man->SetDefaultStorage("raw://");
man->SetRun(281750);
man->SetMaxDate(TTimeStamp(2017,11,7,0,0,0));
man->Get("GRP/CTP/Aliases");
```
Since `SetRun` does caching, it is not possible to run `SetMaxDate` after it. If you now call `SetMaxDate` after the run is set, we exit with `AliFatal`. Please tell me if this is an acceptable behavior.

### 3: MaxDate before run number (OK)
```
man = AliCDBManager::Instance();
man->SetDefaultStorage("raw://");
man->SetMaxDate(TTimeStamp(2017,11,7,0,0,0));
man->SetRun(281750);
man->Get("GRP/CTP/Aliases");
```
Correctly returns `v2` (and always will).

### 4: MaxDate for each specific storage (OK, fixed)
```
man = AliCDBManager::Instance();
man->SetDefaultStorage("alien://folder=/alice/data/2016/OCDB");
man->SetSpecificStorage("GRP/CTP/Aliases","alien://folder=/alice/data/2017/OCDB");
st = man->GetStorage("alien://folder=/alice/data/2017/OCDB");
st->SetMaxDate(TTimeStamp(2017,11,7,0,0,0));
man->SetRun(281750);
man->Get("GRP/CTP/Aliases");
```
Note that it is now possible to invoke `SetMaxDate`, both from the manager and from the storage, with `TTimeStamp` as an argument (no need to run `GetSec()` anymore).

### 5: Get storage manually and set date after (OK, fixed)
```
man = AliCDBManager::Instance();
man->SetDefaultStorage("alien://folder=/alice/data/2016/OCDB");
man->SetSpecificStorage("GRP/CTP/Aliases","alien://folder=/alice/data/2017/OCDB");
stor = (AliCDBStorage*)man->GetActiveStorages()->At(1);
stor->SetMaxDate(TTimeStamp(2017,11,7,0,0,0));
man->SetRun(281750);
man->Get("GRP/CTP/Aliases");
```
This now works correctly (returns `v2`). The rule of running `SetMaxDate` before `SetRun` still applies. In case you accidentally swap them, an `AliFatal` is triggered.

dberzano commented:
@fprino: if @shahor02 agrees with this PR, can we proceed with tagging?

alibuild commented:
f2e615a3d8917656883a5dbf76a57ee5d042e799: tests OK, approval required for merging: 1 of @jgrosseo (Jan Fiete Grosse-Oetringhaus), @chiarazampolli (Chiara Zampolli), @qgp (Jochen Klein), @dberzano (Dario Berzano), @sawenzel (Sandro Christian Wenzel), @ktf (Giulio Eulisse), @shahor02 (Ruben Shahoyan)

_Comment with `+1` to merge. **Please comment on the pull request: [click here](https://github.com/alisw/AliRoot/pull/486) and comment at the bottom of the page.**_

shahor02 commented:
@dberzano : thanks! All mods are ok, including fatal on wrong setting of the timestamp

fprino commented:
Hi Dario,

thanks for these fixes.
For the new AliRoot tag we need to check with Marian the situation with the fixed in the TPC digitiser.
Since he did not open any Jira of porting requests I'd assume that he has still some pending changes.
I'll ping him now via email.

Cheers, Francesco

miranov25 commented:
Hallo Francesco.

* I made porting request  for ion tail digitizer and tracker for MC  in the morning:
  * https://alice.its.cern.ch/jira/browse/ALIROOT-7592
  * I assigned the request to Dario - forgetting to add you and Michael as watcher
* Code is already in master:
  * https://github.com/alisw/AliRoot/pull/485

dberzano commented:
+1 and thanks for your comments!
